docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+address of the implicit object parameter of member functions. Currently that is
+the only place with such an implicit affordance. It is designed to be
+syntactically sound to extend to other parameters, but currently that is not
+planned.
+
+### Dereferencing customization
+
+Carbon should support user-defined pointer-like types such as _smart pointers_
+using a similar pattern as operator overloading or other expression syntax. That
+is, it should rewrite the expression into a member function call on an
+interface. Types can then implement this interface to expose pointer-like
+_user-defined dereference_ syntax.
+
+The interface might look like:
+
+```carbon
+interface Pointer {
+  let ValueT:! Type;
+  fn Dereference[self: Self]() -> ValueT*;
+}
+```
+
+Here is an example using a hypothetical `TaggedPtr` that carries some extra
+integer tag next to the pointer it emulates:
+
+```carbon
+class TaggedPtr(T:! Type) {
+  var tag: Int32;
+  var ptr: T*;
+}
+external impl [T:! Type] TaggedPtr(T) as Pointer {
+  let ValueT:$ T;
+  fn Dereference[self: Self]() -> T* { return self.ptr; }
+}
+
+fn Test(arg: TaggedPtr(T), dest: TaggedPtr(TaggedPtr(T))) {
+  **dest = *arg;
+  *dest = arg;
+}
+```
+
+There is one tricky aspect of this. The function in the interface which
+implements a pointer-like dereference must return a raw pointer which the
+language then actually dereferences to form a reference expression similar to
+that formed by `var` declarations. This interface is implemented for normal
+pointers as a no-op:
+
+```carbon
+impl [T:! Type] T* as Pointer {
+  let ValueT:$ Type = T;
+  fn Dereference[self: Self]() -> T* { return self; }
+}
+```
+
+Dereference expressions such as `*x` are syntactically rewritten to use this
+interface to get a raw pointer and then that raw pointer is dereferenced. If we
+imagine this language level dereference to form a reference expression as a
+unary `deref` operator, then `(*x)` becomes
+`(deref (x.(Pointer.Dereference)()))`.
+
+Carbon will also use a simple syntactic rewrite for implementing `x->Method()`
+as `(*x).Method()` without separate or different customization.
+
+## `const`-qualified types
+
+Carbon provides the ability to qualify a type `T` with the keyword `const` to
+get a `const`-qualified type: `const T`. This is exclusively an API-subsetting
+feature in Carbon -- for more fundamentally "immutable" use cases, value
+expressions and bindings should be used instead. Pointers to `const`-qualified
+types in Carbon provide access to an object with an API subset that can help
+model important requirements like ensuring usage is exclusively by way of a
+_thread-safe_ interface subset of an otherwise _thread-compatible_ type.
+
+The `const T` type has the same representation as `T` with the same field names,
+but all of its field types are also `const`-qualified. Other than fields, all
+other members `T` are also members of `const T`, and impl lookup ignores the
```

Comment:
Wouldn't `addr` methods not be members of `const T` (unless the `self` parameter has type `const Self*`)? I guess that raises questions about how implementations would work. Maybe state that while the `addr` methods would still be members, calling them on a `const T` would be an error if they take a `T*`, but calling non-`addr` methods would generally be fine?